### DISCUSSION TITLE
Running stb_voxel_render on a raspberry pi (OpenGL ES 3.1)

### DISCUSSION THREAD

**Discussion by oakes** [nothings/stb#1355] [Q&A] [answered]
I have a small project that uses stb_voxel_render and am trying to get it running on a raspberry pi 4, which only supports OpenGL ES. I tried just naively changing the `STBVOX_SHADER_VERSION` to use `310 es` but i get the following shader error: `illegal use of reserved word 'usamplerBuffer'`. I guess that type isn't available in this version of GLSL but i do see it listed in the [opengl es 3.0 spec](https://registry.khronos.org/OpenGL/specs/es/3.0/GLSL_ES_Specification_3.00.pdf). I'm not asking for any free development of course, just curious if this is worth digging into or if it would be a non-trivial change.

**Comment by mmozeiko** [comment, marked as answer]
GLSL reserves bunch of keywords for future use. That does not mean their functionality is available.

Texture buffer with usamplerBuffer on GLES is available only in 3.2 version, or when OES_texture_buffer extension is present. See https://en.wikipedia.org/wiki/OpenGL_ES#OpenGL_ES_3.2

In GLES 3.1 (which Pi4 supports) you can use SSBO (shader storage buffer object) for arbitrary buffer lookups in shader - texture buffer can be easily swapped to SSBO. But if you insist on using 3.0 version then you'll need to change code to use regular texture lookups for arbitrary loads.

> on a raspberry pi 4, which only supports OpenGL ES

Raspberry supports desktop OpenGL too, but only version 2.1 (where texture buffer is not available). Here's GL version info on Pi4: https://gist.githubusercontent.com/mmozeiko/8eb6728197099f8f6722b4808964ef43/raw/3a2fa0633abeec5907b1493622d5d17aa24e6a91/info.txt

But you can see ARB_shader_storage_buffer_object available there - so if you use desktop GL version, you should be able to use SSBO there too.
